### ISSUE TITLE
User-oriented demo with estimation subpackage

### ISSUE THREAD

**Issue by aimalz** [LSSTDESC/rail_attic#51]
The current example code for `rail.estimation.Estimator() `and its subclasses is in a script ([here](https://github.com/LSSTDESC/RAIL/blob/master/examples/main.py)), whereas the example code for rail.creation ([here](https://github.com/LSSTDESC/RAIL/blob/issue/7/creation-baseline/examples/creation/basic-creation-demo.ipynb) and [here](https://github.com/LSSTDESC/RAIL/blob/issue/7/creation-baseline/examples/creation/degradation-demo.ipynb)) and rail.evaluation ([here](https://github.com/LSSTDESC/RAIL/blob/issue/4/evaluation-baseline/rail/evaluation/dc1_check.ipynb) and [here](https://github.com/LSSTDESC/RAIL/blob/issue/4/evaluation-baseline/rail/evaluation/demo.ipynb)) are more interactive Jupyter notebooks. This issue is for migrating the content of the rail.estimation demo script into a notebook, while also test driving the intuitiveness of the config dict/file structure and i/o interface from both the perspectives of someone seeking to use an estimator to obtain some photo-z posteriors and of us as experimenters seeking to benchmark many estimators under multiple sets of input data and configs.

**Comment by sylvielsstfr** [comment]
Delight has a kind of tutorial mode in which it generates its mock fluxes in bands for galaixes at random redshift.

Its usefull to have for debugging at expert level, not for standard user.

Do we want to have this mode to be activated from RAIL ?

**Comment by johannct** [comment]
How would you "activate" it? Can you provide an example of how to call this mode and use it?

**Comment by sylvielsstfr** [comment]
I have a flag in the yaml config file which call this tutorial mode

https://github.com/LSSTDESC/RAIL/blob/issue/49/delight/examples/configs/delightPZ.yaml.

For the moment the python file 
https://github.com/LSSTDESC/RAIL/blob/issue/49/delight/rail/estimation/algos/delightPZ.py

works only in that mode. As soon as I can validate the analysis of Tutorial mode provided by Boris through, I can set this tutorial mode to False and read the true training/evaluation I will start to inject DC2 fluxes inside Delight.

But before I will need to be sure I can use the redshift grid proposed by RAIL inside Delight.

I let you decide at which level of the code development you want to control the code by a pull request.
I am not sure this merge now is usefull because RAIL users would need to install my version of Delight.

But we could plan a hack session on this (wednesday at 2P-3PM I have a regular meeting with my lab team, so I am not free).

**Comment by sschmidt23** [comment]
@johannct reminded me that we should include an explanation of saving/loading pretrained models, so I'm mentioning here as a to do item:

- [ ] include demo of storing/loading pretrained model

**Comment by jbkalmbach** [comment]
Went through Sam's notebook last night on branch `issue/48/fileIO` and provided comments. Thought they might be useful in case other people have feedback as well or are interested in the notebook.

# Feedback for Sam
Overall, this is a really helpful notebook. I do feel know like I could start using `rail.estimation` for my photo-z estimation needs. It will definitely be even better as we continue to add packages. Here are my more specific comments on the notebook just written as I was going through.

* Initial activation of cell says all methods are `not installed`.  Installing RAIL didn’t quite work. I ended up having to just add RAIL to my path. Also `qp` isn’t installed as a requirement so should mention the need to install it somewhere.
* Some subheadings in first chunk of explainer text would help organize
* When describing `simpleNN` there is a typo ‘…training data, _then_ assigns…’
* When describing yaml file it looks like `outpath` and `outpath_format` are on the same line as `configpath`
* When describing `nn_dict` typo ‘…on _which_ the PDF …’
* Need `qp` installed to run `rail.estimation.algos.sklearn_nn.simpleNN`
* Would be nice if there were functions to set up the `run_params` dicts for you. That way you could look at the docstring to see all the options when using this.
* Got this warning:
```
/usr/local/py/envs/desc/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:500: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)
```
Is this normal? Should have a note saying to expect this if this is normal. But it didn’t stop the simpleNN training.
* FlexZBoost took 1.5 hours for me. Is that normal?

**Comment by sschmidt23** [comment]
Very useful feedback, thanks Bryce, I will work on implementing some of these changes to make things more clear.

1.5 hours?  FlexZBoost training takes 4-10 minutes on the 10,000 galaxies depending on whether I use my iMac, laptop, or do it on cori (it tests a series of values for both bump thresh and sharpening that adds to the run time).  But 1.5 hours is well above what I would expect.  What system were you running on?  I wonder if it could be related to the specific version of xgboost installed?  I think there was an update to the package recently.  I will look into that, but @jbkalmbach can you tell me what version of xgboost you have installed on your system?

**Comment by sschmidt23** [comment]
Chatting with Bryce, it sounds like the slowness is an xgboost issue, I forgot that when I tested on cori I downgraded to an older version (v0.90) after getting some errors.  I will contact the flexcode developers to ask if they have any updates on xgboost dependencies.

**Comment by sschmidt23** [comment]
I have a demo for estimation that was merged in via PR #60 but I will leave this open in case we want to make any improvements to the notebook, and because it will need to be revised once qp is updated with an append functionality and such.

**Comment by aimalz** [comment]
I think it's okay to close this for now and make a new issue for updating the demos when the relevant qp functionality is ready, since we'll probably want to update other stuff then, too.
